File: src/RuntimeStateStore.cpp
#include "RuntimeStateStore.h"

#include <ArduinoJson.h>

#include "Logger.h"

bool RuntimeStateStore::begin(
    fs::FS& fs,
    LayoutRuntime& runtime) {
  _fs = &fs;
  _runtime = &runtime;

  FileStore files(fs);
  return files.ensureParent("/state/runtime-state.json");
}

bool RuntimeStateStore::load(const char* path) {
  if (!_fs || !_runtime) return false;

  FileStore files(*_fs);
  File file = files.openRead(path);

  if (!file) {
    Logger::info("No saved runtime state; defaults remain active");
    return true;
  }

  JsonDocument doc;
  const DeserializationError error = deserializeJson(doc, file);
  file.close();

  if (error) {
    Logger::warn(String("Runtime state parse failed: ") + error.c_str());
    return false;
  }

  // v3: authoritative physical runtime state only.
  JsonObjectConst basic = doc["basicAccessories"];
  for (JsonPairConst pair : basic) {
    const uint16_t address = static_cast<uint16_t>(atoi(pair.key().c_str()));
    if (!address) continue;
    _runtime->setAccessory(address, pair.value().as<bool>());
  }

  JsonObjectConst extended = doc["extendedAccessories"];
  for (JsonPairConst pair : extended) {
    const uint16_t address = static_cast<uint16_t>(atoi(pair.key().c_str()));
    const int aspect = pair.value().as<int>();
    if (!address || aspect < 0 || aspect > 255) continue;
    _runtime->setSignal(address, aspect);
  }

  JsonObjectConst vpins = doc["vpins"];
  for (JsonPairConst pair : vpins) {
    const uint16_t address = static_cast<uint16_t>(atoi(pair.key().c_str()));
    if (!address) continue;
    _runtime->setVPin(address, pair.value().as<bool>());
  }

  // Read-only migration from old semantic v2 storage. Values are accepted
  // only when they map to the CURRENT layout topology, so protocol changes do
  // not leak stale state across Basic <-> Extended.
  JsonObjectConst legacy = doc["accessories"];
  for (JsonPairConst pair : legacy) {
    JsonObjectConst value = pair.value().as<JsonObjectConst>();
    const char* key = pair.key().c_str();

    if (strncmp(key, "accessory:", 10) == 0) {
      _runtime->setAccessory(
          static_cast<uint16_t>(atoi(key + 10)),
          value["active"] | false);
    } else if (strncmp(key, "signal:", 7) == 0) {
      if (!value["aspect"].isNull()) {
        const int aspect = value["aspect"].as<int>();
        if (aspect >= 0 && aspect <= 255) {
          _runtime->setSignal(
              static_cast<uint16_t>(atoi(key + 7)),
              aspect);
        }
      }
    } else if (strncmp(key, "vpin:", 5) == 0) {
      _runtime->setVPin(
          static_cast<uint16_t>(atoi(key + 5)),
          value["active"] | false);
    } else if (strncmp(key, "turnout:", 8) == 0) {
      const uint16_t address = static_cast<uint16_t>(atoi(key + 8));
      RuntimeAccessory* turnout =
          _runtime->findAccessory(RuntimeAccessoryKind::Turnout, address);

      if (!turnout || turnout->turnoutExtended) continue;

      const bool logicalClosed = value["closed"] | false;
      const bool physicalValue =
          logicalClosed
              ? turnout->closedValue
              : !turnout->closedValue;

      _runtime->setTurnout(address, physicalValue);
    }
  }

  // Sensor state is intentionally never restored. DCC-EX <Q> is authoritative.

  JsonObjectConst blocks = doc["blocks"];
  for (JsonPairConst pair : blocks) {
    const long blockIdValue = atol(pair.key().c_str());
    if (blockIdValue <= 0 || blockIdValue > 0xffff) continue;

    JsonObjectConst value = pair.value().as<JsonObjectConst>();

    const String locoId = value["locoId"].isNull()
        ? String()
        : String(value["locoId"].as<const char*>());

    const long locoAddressValue = value["locoAddress"] | 0L;
    const uint16_t locoAddress =
        locoAddressValue > 0 && locoAddressValue <= 10239
            ? static_cast<uint16_t>(locoAddressValue)
            : 0;

    _runtime->setBlock(
        static_cast<uint16_t>(blockIdValue),
        locoId,
        locoAddress);
  }

  Logger::info("Physical runtime state restored");
  return true;
}

bool RuntimeStateStore::save(const char* path) {
  if (!_fs || !_runtime) return false;

  JsonDocument doc;
  doc["version"] = 3;
  doc["savedAtMs"] = millis();

  JsonObject basic = doc["basicAccessories"].to<JsonObject>();
  for (const auto& item : _runtime->basicAccessories()) {
    basic[String(item.address)] = item.active;
  }

  JsonObject extended = doc["extendedAccessories"].to<JsonObject>();
  for (const auto& item : _runtime->extendedAccessories()) {
    extended[String(item.address)] = item.aspect;
  }

  JsonObject vpins = doc["vpins"].to<JsonObject>();
  for (const auto& item : _runtime->vpinStates()) {
    vpins[String(item.address)] = item.active;
  }

  JsonObject blocks = doc["blocks"].to<JsonObject>();
  for (const auto& block : _runtime->blocks()) {
    if (!block.occupied()) continue;

    JsonObject state = blocks[String(block.id)].to<JsonObject>();
    if (block.locoId.isEmpty()) state["locoId"] = nullptr;
    else state["locoId"] = block.locoId;

    if (block.locoAddress > 0) state["locoAddress"] = block.locoAddress;
  }

  FileStore files(*_fs);
  if (!files.saveJson(path, doc)) {
    Logger::error("Runtime state atomic save failed");
    return false;
  }

  Logger::info("Physical runtime state saved on POWER OFF");
  return true;
}

File: desktop/DCCExpressHub.Net/Web/RuntimeStateStore.cs
using System.Text.Json;

namespace DCCExpressHub.Net.Web;

public sealed class RuntimeStateStore
{
    readonly LayoutRuntime _runtime;
    readonly string _path;
    readonly ILogger<RuntimeStateStore> _log;
    static readonly JsonSerializerOptions Json = new(JsonSerializerDefaults.Web) { WriteIndented = true };

    public RuntimeStateStore(LayoutRuntime runtime, IWebHostEnvironment env, ILogger<RuntimeStateStore> log)
    {
        _runtime = runtime;
        _log = log;
        _path = Path.Combine(env.ContentRootPath, "data", "state", "runtime-state.json");
        Directory.CreateDirectory(Path.GetDirectoryName(_path)!);
    }

    public async Task<bool> LoadAsync()
    {
        if (!File.Exists(_path))
        {
            _log.LogInformation("No saved runtime state; defaults remain active");
            return true;
        }

        try
        {
            using var doc = JsonDocument.Parse(await File.ReadAllTextAsync(_path));
            var root = doc.RootElement;

            // v3 physical runtime format.
            if (root.TryGetProperty("basicAccessories", out var basic) && basic.ValueKind == JsonValueKind.Object)
            {
                foreach (var pair in basic.EnumerateObject())
                {
                    if (!ushort.TryParse(pair.Name, out var address))
                        continue;

                    bool active = pair.Value.ValueKind == JsonValueKind.True ||
                                  (pair.Value.ValueKind == JsonValueKind.Number && pair.Value.TryGetInt32(out var n) && n != 0);

                    _runtime.SetAccessory(address, active);
                }
            }

            if (root.TryGetProperty("extendedAccessories", out var extended) && extended.ValueKind == JsonValueKind.Object)
            {
                foreach (var pair in extended.EnumerateObject())
                {
                    if (!ushort.TryParse(pair.Name, out var address))
                        continue;

                    if (!pair.Value.TryGetInt32(out var aspect) || aspect is < 0 or > 255)
                        continue;

                    _runtime.SetSignal(address, aspect);
                }
            }

            if (root.TryGetProperty("vpins", out var vpins) && vpins.ValueKind == JsonValueKind.Object)
            {
                foreach (var pair in vpins.EnumerateObject())
                {
                    if (!ushort.TryParse(pair.Name, out var address))
                        continue;

                    bool active = pair.Value.ValueKind == JsonValueKind.True ||
                                  (pair.Value.ValueKind == JsonValueKind.Number && pair.Value.TryGetInt32(out var n) && n != 0);

                    _runtime.SetVPin(address, active);
                }
            }

            // Read-only migration support for the old semantic v2 format.
            // Only values that map cleanly to current physical topology survive.
            if (root.TryGetProperty("accessories", out var legacyAccessories) && legacyAccessories.ValueKind == JsonValueKind.Object)
            {
                foreach (var pair in legacyAccessories.EnumerateObject())
                {
                    var key = pair.Name;
                    var value = pair.Value;

                    if (key.StartsWith("accessory:", StringComparison.Ordinal) &&
                        ushort.TryParse(key[10..], out var basicAddress))
                    {
                        _runtime.SetAccessory(
                            basicAddress,
                            value.TryGetProperty("active", out var active) && active.ValueKind == JsonValueKind.True);
                    }
                    else if (key.StartsWith("signal:", StringComparison.Ordinal) &&
                             ushort.TryParse(key[7..], out var extendedAddress) &&
                             value.TryGetProperty("aspect", out var a) &&
                             a.ValueKind == JsonValueKind.Number &&
                             a.TryGetInt32(out var aspect) &&
                             aspect is >= 0 and <= 255)
                    {
                        // This succeeds only if the CURRENT layout says the address
                        // is Extended. A former basic signal cannot leak into it.
                        _runtime.SetSignal(extendedAddress, aspect);
                    }
                    else if (key.StartsWith("vpin:", StringComparison.Ordinal) &&
                             ushort.TryParse(key[5..], out var vp))
                    {
                        _runtime.SetVPin(
                            vp,
                            value.TryGetProperty("active", out var active) && active.ValueKind == JsonValueKind.True);
                    }
                    else if (key.StartsWith("turnout:", StringComparison.Ordinal) &&
                             ushort.TryParse(key[8..], out var turnoutAddress) &&
                             value.TryGetProperty("closed", out var c) &&
                             c.ValueKind is JsonValueKind.True or JsonValueKind.False)
                    {
                        var turnout = _runtime.FindAccessory(RuntimeAccessoryKind.Turnout, turnoutAddress);
                        if (turnout is null || turnout.TurnoutExtended)
                            continue;

                        bool logicalClosed = c.GetBoolean();
                        bool physical = logicalClosed ? turnout.ClosedValue : !turnout.ClosedValue;
                        _runtime.SetTurnout(turnoutAddress, physical);
                    }
                }
            }

            // Sensor state is intentionally NOT restored. DCC-EX <Q> is the
            // authoritative source and startup/reconnect requests it.

            if (root.TryGetProperty("blocks", out var blocks) && blocks.ValueKind == JsonValueKind.Object)
            {
                foreach (var pair in blocks.EnumerateObject())
                {
                    if (!ushort.TryParse(pair.Name, out var blockId) || blockId == 0)
                        continue;

                    var value = pair.Value;
                    string locoId = value.TryGetProperty("locoId", out var li) && li.ValueKind == JsonValueKind.String
                        ? li.GetString() ?? ""
                        : "";

                    ushort locoAddress = 0;
                    if (value.TryGetProperty("locoAddress", out var la) &&
                        la.TryGetInt32(out var n) &&
                        n is > 0 and <= 10239)
                    {
                        locoAddress = (ushort)n;
                    }

                    _runtime.SetBlock(blockId, locoId, locoAddress);
                }
            }

            _log.LogInformation("Physical runtime state restored");
            return true;
        }
        catch (Exception ex)
        {
            _log.LogWarning(ex, "Runtime state parse failed");
            return false;
        }
    }

    public async Task<bool> SaveAsync()
    {
        try
        {
            var basic = _runtime.BasicAccessoryStatesForPersistence()
                .ToDictionary(x => x.Key.ToString(), x => x.Value);

            var extended = _runtime.ExtendedAccessoryStatesForPersistence()
                .ToDictionary(x => x.Key.ToString(), x => x.Value);

            var vpins = _runtime.VPinStatesForPersistence()
                .ToDictionary(x => x.Key.ToString(), x => x.Value);

            var blocks = new Dictionary<string, object>();
            foreach (var block in _runtime.BlocksForPersistence())
            {
                if (string.IsNullOrEmpty(block.LocoId) && block.LocoAddress == 0)
                    continue;

                blocks[block.Id.ToString()] = new
                {
                    locoId = string.IsNullOrEmpty(block.LocoId) ? null : block.LocoId,
                    locoAddress = block.LocoAddress > 0 ? (ushort?)block.LocoAddress : null
                };
            }

            var body = JsonSerializer.Serialize(new
            {
                version = 3,
                savedAtMs = Environment.TickCount64,
                basicAccessories = basic,
                extendedAccessories = extended,
                vpins,
                blocks
            }, Json);

            var tmp = _path + ".tmp";
            await File.WriteAllTextAsync(tmp, body);
            File.Move(tmp, _path, true);

            _log.LogInformation("Physical runtime state saved on POWER OFF");
            return true;
        }
        catch (Exception ex)
        {
            _log.LogError(ex, "Runtime state atomic save failed");
            return false;
        }
    }
}
